chore: add `unicode` task, remove python script

malobre opened the pull request:
## Summary
- Closes #1930.

## Test Plan
- [x] `cargo check`
- [x] `cargo fmt`
- [x] `cargo test --package rslint_lexer`
- [x] Compare generated tables: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=912a6a48a2c98a9cc6b338e9888cda6e)

MichaReiser commented on xtask/src/unicode.rs:
```diff
@@ -0,0 +1,112 @@
+		}
+	};
+
+	let pretty = prettyplease::unparse(&syn::parse2::<syn::File>(tokens)?);
```
You can use the `reformat` function that all other codegens are using (avoids yet another dependency that)

https://github.com/rome/tools/blob/7f555aa275b3629cc6cbc9a74f7dc2e9b88e6f68/xtask/src/lib.rs#L52

malobre replied:
I thought `prettyplease` would be better because it wouldn't rely on invoking `rustfmt` binary (feels hacky), plus it only depends on `syn` and `proc-macro2`, according to the readme it's also faster than `rustfmt`.

However I understand this might bloat the PR and would be better discussed in its own issue / PR.

malobre commented:
I cleared the code a bit and added comments.

malobre commented:
There seems to be some discrepancies in the generated code, I'm looking into it.

malobre commented:
I've updated the code to match the behaviour of the python script regarding the collapsing of sequential ranges.
The generated arrays are now [identical](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=782b8c0a596346c881cb4316fc90b0c5).

malobre commented:
There, I'm happy with the code. No more changes I swear ! :laughing:

malobre commented:
Rebased on `main`.

MichaReiser commented on Cargo.lock:
```diff
@@ -1087,6 +1087,16 @@ dependencies = [
  "plotters-backend",
 ]
 
+[[package]]
+name = "prettyplease"
+version = "0.1.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "3f3ebd569945e276e2c8000677484f50f5c242b3495929ae1c7d42d89778939a"
+dependencies = [
+ "proc-macro2",
+ "syn",
+]
+
```
I would favour it if we swap to `prettyplease` as part of another PR.

MichaReiser commented on crates/rslint_lexer/src/tables.rs:
```diff
@@ -1,468 +1,1427 @@
+fn bsearch_range_table(c: char, r: &[(char, char)]) -> bool {
     use core::cmp::Ordering::{Equal, Less, Greater};
-    r.binary_search_by(|&(lo,hi)| {
-        // Because ASCII ranges are at the start of the tables, a search for an
-        // ASCII char will involve more `Greater` results (i.e. the `(lo,hi)`
-        // table entry is greater than `c`) than `Less` results. And given that
-        // ASCII chars are so common, it makes sense to favor them. Therefore,
-        // the `Greater` case is tested for before the `Less` case.
```
Should we keep this comment?

MichaReiser commented on crates/rslint_lexer/src/tables.rs:
```diff
@@ -1,468 +1,1427 @@
+        ('𬺰', '𮯠'),
+        ('丽', '𪘀'),
+        ('𰀀', '𱍊'),
+        ('\u{e0100}', '\u{e01ef}'),
```
What's the reason that this entry uses unicode escape but you migrated all other to unicode characters

malobre replied:
This is not by choice, I wondered about this too and I'm not sure why but only some char are escaped. This might be related to https://github.com/rust-lang/rust/issues/60495.

MichaReiser commented on xtask/src/unicode.rs:
```diff
@@ -0,0 +1,168 @@
+    /// Retrieve properties from cache, see [`Self::save_cache`].
+    /// # Errors
+    /// Return an error if the cache file is missing.
+    pub fn from_cache() -> Result<Self> {
+        let raw = std::fs::read_to_string(Self::path())?;
+
+        Ok(Self { raw })
+    }
+
+    /// Retrieve properties from the unicode website.
+    /// # Errors
+    /// Return an error if the HTTP request fails.
+    pub fn fetch() -> Result<Self> {
+        let raw = ureq::get("http://www.unicode.org/Public/UNIDATA/DerivedCoreProperties.txt")
+            .call()?
+            .into_string()?;
+
+        Ok(Self { raw })
+    }
+
+    /// Save the properties on disk for later usage, see [`Self::from_cache`].
+    /// # Errors
+    /// Return an error if saving to disk fails.
+    pub fn save_cache(&self) -> Result<()> {
+        std::fs::write(Self::path(), &self.raw)?;
+
+        Ok(())
+    }
```
I would make these private methods and expose a single public method that retrieves the properties from cache or downloads them if missing. 

We may want to print to stdout in the case the script uses a cached file. Otherwise someone may run the script to update the unicode table but see no changes and wonder why.

MichaReiser commented on xtask/src/unicode.rs:
```diff
@@ -0,0 +1,168 @@
+        fn bsearch_range_table(c: char, r: &[(char,char)]) -> bool {
+            use core::cmp::Ordering::{Equal, Less, Greater};
+            r.binary_search_by(|&(lo,hi)| {
+                // Because ASCII ranges are at the start of the tables, a search for an
```
Ah, the comment is still here. That's good enough.

MichaReiser commented on xtask/src/unicode.rs:
```diff
@@ -0,0 +1,168 @@
+                ];
+
+                pub fn #fn_ident(c: char) -> bool {
+                    super::bsearch_range_table(c, #table_ident)
```
The implementation uses a binary search. How does this script guarantee that the properties are sorted. Is this guaranteed by the unicode table? Would it make sense to re-sort it or at least assert that the entries are sorted?

malobre replied:
The unicode table is sorted, but I couldn't find any mention of it in the [Format Convention](https://www.unicode.org/reports/tr44/#Format_Conventions).
I'll add a call to `slice::sort_unstable`.

MichaReiser reviewed (commented):
I'm not familiar with the python script but only have some overall comments:

* Did you run your new script. Does it generate the exact same output?

malobre commented:
I have addressed all comments (except for the `char` escaping situation).
@MichaReiser The generated tables are identical: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=912a6a48a2c98a9cc6b338e9888cda6e

MichaReiser reviewed (approved):
Excellent work! Thank you so much for working on this.
